### PR TITLE
Pcap read directories recursively v5

### PR DESCRIPTION
**Pull request by jamesdutrisac**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [ x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [ x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [ x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/2363

Describe changes:
- Added ability to recursively read pcap directories
- src/suricata.c: addition of new command line parameter
    --pcap-file-recursive
- src/source-pcap-file.c: parsing of the command line argument
- src/source-pcap-file-directory-helper.h: two thread vars tracking directory depth and should recurse
- src/util-path.c / src/util-path.h: added 4 functions SCIsRegularFile and SCIsRegularDirectory to provide OS independent
    file / directory info.
    SCRealPath is an OS independent wrapper for realpath
    PathJoin to manage path resolution logic
- src/util-error.c / src/util-error.h: 
   Added new warning code "SC_WARN_PATH_READ_ERROR"

- This updates version 4 (https://github.com/OISF/suricata/pull/5157) with the following a bug fix where the variable used to track the depth was overflowing before the exit condition was met.  Error checking on opening directories.  And an update to the comments.

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

#suricata-verify-pr:
#suricata-verify-repo:
#suricata-verify-branch:
#suricata-update-pr:
#suricata-update-repo:
#suricata-update-branch:
#libhtp-pr:
#libhtp-repo:
#libhtp-branch:

### CODE REVIEW

File: src/util-path.c [hunk 1]
```diff
@@ -66,6 +66,35 @@ int PathIsRelative(const char *path)
...
+        SCLogError(SC_ERR_SPRINTF, "Could not join filename to path");
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+    char *tmpBuf = SCRealPath(out_buf, NULL);
+    if (tmpBuf == NULL) {
+        SCLogError(SC_ERR_SPRINTF, "Error resolving path: %s", strerror(errno));
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+    memset(out_buf, 0, buf_len);
+    snprintf(out_buf, max_path_len, "%s", tmpBuf);
```
Comment (victorjulien) [inline src/util-path.c:93]: Can we use `strlcpy` here?

File: src/util-path.c [hunk 2]
```diff
@@ -66,6 +66,35 @@ int PathIsRelative(const char *path)
...
+TmEcode PathJoin (char *out_buf, uint16_t buf_len, const char *const dir, const char *const fname)
+{
+    memset(out_buf, 0, buf_len);
+    uint16_t max_path_len = (buf_len > PATH_MAX) ? PATH_MAX : buf_len;
+    int bytes_written = snprintf(out_buf, max_path_len, "%s/%s", dir, fname);
+    if (bytes_written <= 0) {
+        SCLogError(SC_ERR_SPRINTF, "Could not join filename to path");
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+    char *tmpBuf = SCRealPath(out_buf, NULL);
```
Comment (victorjulien) [inline src/util-path.c:87]: nit: style `tmp_buf`

File: src/util-path.c [hunk 3]
```diff
@@ -66,6 +66,35 @@ int PathIsRelative(const char *path)
...
+ * @param buf_len length of output buffer
+ * @param dir the directory
+ * @param fname the filename
+ * @return TM_ECODE_OK on success.  On failure TM_ECODE_FAILED is returned.
+ */
+TmEcode PathJoin (char *out_buf, uint16_t buf_len, const char *const dir, const char *const fname)
+{
+    memset(out_buf, 0, buf_len);
+    uint16_t max_path_len = (buf_len > PATH_MAX) ? PATH_MAX : buf_len;
+    int bytes_written = snprintf(out_buf, max_path_len, "%s/%s", dir, fname);
```
Comment (victorjulien) [inline src/util-path.c:82]: Can be done later, but I guess here we would use `\` for Windows?

File: src/source-pcap-file.c [hunk 4]
```diff
@@ -292,13 +292,26 @@ TmEcode ReceivePcapFileThreadInit(ThreadVars *tv, const void *initdata, void **d
             CleanupPcapFileThreadVars(ptv);
             SCReturnInt(TM_ECODE_OK);
         }
+        pv->cur_dir_depth = 0;
+
+        int should_recurse = 0;
+        pv->should_recurse = false;
+        if (ConfGetBool("pcap-file.recursive", &should_recurse) == 1) {
+            pv->should_recurse = should_recurse == 1;
```
Comment (victorjulien) [inline src/source-pcap-file.c:300]: nit: `= true`.

Also, doesn't look like `should_recurse` is used after this, so setting it is unnecessary.

File: src/source-pcap-file-directory-helper.c [hunk 5]
```diff
@@ -314,82 +316,107 @@ TmEcode PcapDirectoryPopulateBuffer(PcapFileDirectoryVars *pv,
...
+             * Only set pv->directory on success */
+            DIR *directory = opendir(next_dir);
+            if (directory == NULL) {
+                SCLogWarning(SC_WARN_PATH_READ_ERROR, "Error opening directory %s.  %s", next_dir, strerror(errno));
+                SCReturnInt(TM_ECODE_FAILED);
+            }
+            pv->directory = directory;
+            pv->filename = next_dir;
+            /* Recursively descend into directory structure */
+            ++pv->cur_dir_depth;
```
Comment (victorjulien) [inline src/source-pcap-file-directory-helper.c:357]: should this guard against overflowing?

File: src/source-pcap-file-directory-helper.c [hunk 6]
```diff
@@ -314,82 +316,107 @@ TmEcode PcapDirectoryPopulateBuffer(PcapFileDirectoryVars *pv,
...
+            PcapDirectoryPopulateBuffer(pv, older_than);
+            --pv->cur_dir_depth;
```
Comment (victorjulien) [inline src/source-pcap-file-directory-helper.c:359]: Can this go below 0? (underflowing)

Guess I'm not following the depth logic yet.

File: src/util-path.c (hunk 3 above)
Comment (jamesdutrisac) [reply src/util-path.c:82]: Good point, may as well do it now.

File: src/source-pcap-file.c (hunk 4 above)
Comment (jamesdutrisac) [reply src/source-pcap-file.c:300]: I'll clean this up.

File: src/util-path.c (hunk 1 above)
Comment (jamesdutrisac) [reply src/util-path.c:93]: Good point.

File: src/source-pcap-file.c (hunk 4 above)
Comment (jamesdutrisac) [reply src/source-pcap-file.c:300]: @victorjulien I took a closer look at this.  The function `ConfGetBool` returns the sets the variable as an int.  However, it's stored in the threadvars as a bool.  So this is being used to convert from the int value to the bool.  It is the same paradigm that's being used below with pcap-file.continuous.  Two suggestions that would be clearer are to do either:
`pv->should_recurse = (should_recurse == 1) ? true : false;`
or
`pv->should_recurse = (should_recurse == 1);`
Do you have a preference?  And should I also go ahead and update pcap-file.continuous?

File: src/source-pcap-file-directory-helper.c (hunk 5 above)
Comment (jamesdutrisac) [reply src/source-pcap-file-directory-helper.c:357]: There is a check when entering a recursive directory to make sure that the traversal is still within the limit: 
`if ((pv->should_recurse == true) && (pv->cur_dir_depth >= max_directory_recursion)`

That check should double as an overflow guard.

File: src/source-pcap-file-directory-helper.c (hunk 6 above)
Comment (jamesdutrisac) [reply src/source-pcap-file-directory-helper.c:359]: There shouldn't be a risk of this wrapping around.  The logic is:
- Recursively descend to the deepest allowable directory to build the directory tree.  This is what `cur_dir_depth` tracks.
- On the return path add each file in the directory to the pending file buffer

**Comment by jamesdutrisac** [comment]
Closing to open v6: https://github.com/OISF/suricata/pull/5211

**Pull request closed without merging**